worker/uniter/runner/runner.go

```diff
@@ -313,35 +317,105 @@ func (runner *runner) runCharmHookWithLocation(hookName, charmLocation string, r
 	return runner.runCharmHookOnLocal(hookName, env, charmLocation)
 }
 
+// loggerAdaptor implements MessageReceiver and
+// sends messages to a logger.
+type loggerAdaptor struct {
+	loggo.Logger
+}
+
+func (l *loggerAdaptor) Messagef(isPrefix bool, message string, args ...interface{}) {
+	l.Debugf(message, args...)
+}
+
+// bufferAdaptor implements MessageReceiver and
+// is used with the out writer from os.Pipe().
+// It allows the hook logger to grab console output
+// as well as passing the output to an action result.
+type bufferAdaptor struct {
+	io.ReadWriter
+	outCopy bytes.Buffer
+}
+
+func (b *bufferAdaptor) Read(p []byte) (n int, err error) {
+	return b.outCopy.Read(p)
+}
+
+func (b *bufferAdaptor) Messagef(isPrefix bool, message string, args ...interface{}) {
+	formattedMessage := message
+	if len(args) > 0 {
+		formattedMessage = fmt.Sprintf(message, args...)
+	}
+	if !isPrefix {
+		formattedMessage += "\n"
+	}
+	b.outCopy.WriteString(formattedMessage)
+}
+
 func (runner *runner) runCharmHookOnRemote(hookName string, env []string, charmLocation string) error {
 	charmDir := runner.paths.GetCharmDir()
 	hook := filepath.Join(charmDir, filepath.Join(charmLocation, hookName))
 
 	var cancel chan struct{}
 	outReader, outWriter, err := os.Pipe()
 	if err != nil {
-		return errors.Errorf("cannot make logging pipe: %v", err)
+		return errors.Errorf("cannot make stdout logging pipe: %v", err)
 	}
 	defer outWriter.Close()
 
-	hookLogger := charmrunner.NewHookLogger(runner.getLogger(hookName), outReader)
-	defer hookLogger.Stop()
-	go hookLogger.Run()
+	actionOut := &bufferAdaptor{ReadWriter: outWriter}
+	hookOutLogger := charmrunner.NewHookLogger(outReader,
+		&loggerAdaptor{runner.getLogger(hookName)},
+		actionOut,
+	)
+	defer hookOutLogger.Stop()
+	go hookOutLogger.Run()
+
+	// When running an action, We capture stdout and stderr
+	// separately to pass back.
+	var actionErr = actionOut
+	_, err = runner.context.ActionData()
+	runningAction := err == nil
+	if runningAction {
+		errReader, errWriter, err := os.Pipe()
+		if err != nil {
+			return errors.Errorf("cannot make stderr logging pipe: %v", err)
+		}
+		defer errWriter.Close()
+
+		actionErr = &bufferAdaptor{ReadWriter: errWriter}
+		hookErrLogger := charmrunner.NewHookLogger(errReader,
+			&loggerAdaptor{runner.getLogger(hookName)},
+			actionErr,
+		)
+		defer hookErrLogger.Stop()
+		go hookErrLogger.Run()
+	}
 
 	executor, err := runner.getRemoteExecutor(runOnRemote)
 	if err != nil {
 		return errors.Trace(err)
 	}
-	_, err = executor(
+	resp, err := executor(
 		ExecParams{
-			Commands:   []string{hook},
-			Env:        env,
-			WorkingDir: charmDir,
-			Stdout:     outWriter,
-			Stderr:     outWriter,
-			Cancel:     cancel,
+			Commands:        []string{hook},
+			Env:             env,
+			WorkingDir:      charmDir,
+			Stdout:          actionOut,
+			Stderr:          actionErr,
+			Cancel:          cancel,
+			ProcessResponse: true,
```

Comment:
we can just remove `ProcessResponse` from the struct and related code in `getNewRunnerExecutor` now.

Reply:
yep